src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Parser.java

```diff
@@ -0,0 +1,503 @@
+    }
+
+    /*
+     * Newline characters in the returned text are of the \n form.
+     */
+    public Result parse(String eolMarker, String source) throws ParseException {
+        Objects.requireNonNull(eolMarker);
+        Objects.requireNonNull(source);
+        if (!Objects.equals(eolMarker, this.eolMarker)) {
+            if (eolMarker.length() < 1) {
+                throw new IllegalArgumentException();
+            }
+            for (int i = 0; i < eolMarker.length(); i++) {
+                switch (eolMarker.charAt(i)) {
+                    case '\f', '\n', '\r' -> throw new IllegalArgumentException();
+                }
+            }
+            this.eolMarker = eolMarker;
+            // capture the rightmost eolMarker (e.g. "//")
+            // The bellow Pattern.compile should never throw PatternSyntaxException
+            Pattern pattern = Pattern.compile("^(.*)(" + Pattern.quote(eolMarker) + "(\\s*@\\s*\\w+.+?))$");
+            this.markedUpLine = pattern.matcher(""); // reusable matcher
+        }
+
+        tags.clear();
+        regions.clear();
+
+        Queue<Action> actions = new LinkedList<>();
+
+        StyledText text = new StyledText();
+        boolean trailingNewline = source.endsWith("\r") || source.endsWith("\n");
+        int lineStart = 0;
+        List<Tag> previousLineTags = new ArrayList<>();
+        List<Tag> thisLineTags = new ArrayList<>();
+        List<Tag> tempList = new ArrayList<>();
+
+        // while lines could be computed lazily, it would yield more complex code
+        record OffsetAndLine(int offset, String line) { }
+        var offsetAndLines = new LinkedList<OffsetAndLine>();
+        forEachLine(source, (off, line) -> offsetAndLines.add(new OffsetAndLine(off, line)));
+        Iterator<OffsetAndLine> iterator = offsetAndLines.iterator();
+
+        while (iterator.hasNext()) {
+            // There are 3 cases:
+            //   1. The pattern that describes a marked-up line is not matched
+            //   2. While the pattern is matched, the markup is not recognized
+            //   3. Both the pattern is matched and the markup is recognized
+            OffsetAndLine next = iterator.next();
+            String rawLine = next.line();
+            boolean addLineTerminator = iterator.hasNext() || trailingNewline;
+            String line;
+            markedUpLine.reset(rawLine);
+            if (!markedUpLine.matches()) { // (1)
+                line = rawLine + (addLineTerminator ? "\n" : "");
+            } else {
+                String maybeMarkup = markedUpLine.group(3);
+                List<Tag> parsedTags = markupParser.parse(maybeMarkup);
+                for (Tag t : parsedTags) {
+                    t.lineSourceOffset = next.offset;
+                    t.markupLineOffset = markedUpLine.start(3);
+                }
+                thisLineTags.addAll(parsedTags);
+                for (var tagIterator = thisLineTags.iterator(); tagIterator.hasNext(); ) {
```

Comment:
oooh, not wrong; just unexpected mix of styles ;-)

Reply:
Can you be more specific? In what sense is this unexpected?

Reply:
Just the combination of `var`, iterators, and a `for `(;;)` loop, that's all. Not wrong, just uncommon; it took me a double-take to parse it, that's all.   The code is OK, no need to change it.